### PR TITLE
Shader-input parser

### PR DESCRIPTION
**Pull request by MeucheRoume**

### CODE REVIEW

File: post-processing-examples/bichrome.frag [hunk 1]
```diff
@@ -7,23 +7,25 @@ out vec4      out_Color;
...
-uniform float Effect_intensity;
+uniform RgbColor Color1;
+uniform RgbColor Color2;
+uniform float    Effect_intensity; // default 0.5 min -1 max 2
 
 // END DYNAMIC PARAMS
 
 void main()
 {
-    vec3 in_color = image(_uv);
```
Comment (MeucheRoume) [inline post-processing-examples/bichrome.frag:24]: TODO: Don't use the strong types outside of the uniform declarations

File: src/Module_CustomShader/Module_CustomShader.cpp [hunk 2]
```diff
@@ -125,8 +127,134 @@ static void keep_values_of_inputs_that_already_existed_and_destroy_unused_ones(
     }
 }
 
-static auto get_inputs_from_shader_code(std::string_view source_code, DirtyFlag dirty_flag, InputFactory_Ref input_factory)
-    -> std::vector<AnyInput>
+template<typename T>
```
Comment (MeucheRoume) [inline src/Module_CustomShader/Module_CustomShader.cpp:130]: Add documenting comment

File: post-processing-examples/bichrome.frag [hunk 3]
```diff
@@ -7,23 +7,21 @@ out vec4      out_Color;
 
 uniform sampler2D _image;
 
+// #include "_COOL_RES_/shaders/input_definitions.glsl"
 // #include "_ROOT_FOLDER_/shader-lib/image.glsl"
 // #include "_ROOT_FOLDER_/shader-lib/bichrome.glsl"
 
-// BEGIN DYNAMIC PARAMS
+// default 0.5 min -1 max 1
```
Comment (JulesFouchy) [inline post-processing-examples/bichrome.frag:14]: ?

File: post-processing-examples/bichrome.frag [hunk 4]
```diff
@@ -7,23 +7,21 @@ out vec4      out_Color;
...
 
-uniform vec3 Color1;
-uniform vec3 Color2;
-
-uniform float Effect_intensity;
-
-// END DYNAMIC PARAMS
+INPUT RgbColor Color1;           // default (0.5, 0.5, 0.5) hdr
+INPUT RgbColor Color2;           // hdr default (0.3, 0.6, 0.2)
+INPUT float    Effect_intensity; // default 0.5 min -1 max 2
```
Comment (JulesFouchy) [inline post-processing-examples/bichrome.frag:18]: Shouldn't all Effect_intensity default to 1?

File: post-processing-examples/bichrome.frag [hunk 5]
```diff
@@ -7,23 +7,21 @@ out vec4      out_Color;
...
 
 void main()
 {
-    vec3 in_color = image(_uv);
+    RgbColor in_color = image(_uv);
```
Comment (JulesFouchy) [inline post-processing-examples/bichrome.frag:22]: Don't use our special types outside of an INPUT declaration

File: post-processing-examples/hue.frag [hunk 6]
```diff
@@ -7,24 +7,21 @@ out vec4      out_Color;
 
 uniform sampler2D _image;
 
+// #include "_COOL_RES_/shaders/input_definitions.glsl"
 // #include "_ROOT_FOLDER_/shader-lib/image.glsl"
 // #include "_ROOT_FOLDER_/shader-lib/hue.glsl"
 
-// BEGIN DYNAMIC PARAMS
+/*truc*/ INPUT Hue Hue_coeficient;
```
Comment (JulesFouchy) [inline post-processing-examples/hue.frag:14]: ?

File: post-processing-examples/psrdnoise.frag [hunk 7]
```diff
@@ -28,7 +28,7 @@
 // can make "flow noise" effects as presented by Perlin and Neyret.
 //
 // vec3 {p}s{r}dnoise(vec2 pos {, vec2 per} {, float rot})
-// "pos" is the input (x,y) coordinate
+// "pos" is the INPUT (x,y) coordinate
```
Comment (JulesFouchy) [inline post-processing-examples/psrdnoise.frag:31]: ?

File: post-processing-examples/psrdnoise.frag [hunk 8]
```diff
@@ -37,7 +37,7 @@
 // The second and third components are the x and y partial derivatives.
 //
 // float {p}s{r}noise(vec2 pos {, vec2 per} {, float rot})
-// "pos" is the input (x,y) coordinate
+// "pos" is the INPUT (x,y) coordinate
```
Comment (JulesFouchy) [inline post-processing-examples/psrdnoise.frag:40]: ?

File: post-processing-examples/psrdnoise.frag [hunk 9]
```diff
@@ -59,7 +59,7 @@
 //
 // TODO: One-pixel wide artefacts used to occur due to precision issues with
 // the gradient indexing. This is specific to this variant of noise, because
-// one axis of the simplex grid is perfectly aligned with the input x axis.
+// one axis of the simplex grid is perfectly aligned with the INPUT x axis.
```
Comment (JulesFouchy) [inline post-processing-examples/psrdnoise.frag:62]: ?

File: src/Module_CustomShader/Module_CustomShader.cpp [hunk 10]
```diff
@@ -1,13 +1,20 @@
...
 #include <Cool/Log/ToUser.h>
+#include <Cool/String/String.h>
+#include <doctest/doctest.h>
 #include <glpp/glpp.hpp>
 #include <ranges>
 #include <sstream>
+#include <type_from_string/type_from_string.hpp>
 
-namespace Lab {
+// Support angle units (turns, degrees, radians)
```
Comment (JulesFouchy) [inline src/Module_CustomShader/Module_CustomShader.cpp:12]: Add a TODO(LD)

File: src/Module_CustomShader/Module_CustomShader.cpp [hunk 11]
```diff
@@ -32,31 +39,31 @@ void Module_CustomShader::imgui_windows(Ui_Ref ui) const
...
 template<typename T>
-static void set_uniform(const Cool::OpenGL::Shader& shader, std::string_view name, const T& value)
+static auto set_uniform(const Cool::OpenGL::Shader& shader, std::string_view name, const T& value) -> void
 {
     shader.set_uniform(name, value);
 }
 
 template<>
-void set_uniform(const Cool::OpenGL::Shader& shader, std::string_view name, const Cool::RgbColor& value)
+auto set_uniform(const Cool::OpenGL::Shader& shader, std::string_view name, const Cool::RgbColor& value) -> void
```
Comment (JulesFouchy) [inline src/Module_CustomShader/Module_CustomShader.cpp:48]: Dont use a trailing return type for void

**Pull request merged**